File: backend/app/schemas/project.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel


class ProjectCreate(BaseModel):
    title: str
    description: Optional[str] = None
    status: Optional[str] = "active"
    due_date: Optional[date] = None
    sort_order: Optional[int] = 0


class ProjectUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    due_date: Optional[date] = None
    sort_order: Optional[int] = None


class TaskInProject(BaseModel):
    id: int
    title: str
    status: str
    priority: int
    due_date: Optional[date] = None
    sort_order: int
    depth: int
    created_at: datetime

    model_config = {"from_attributes": True}


class ProjectResponse(BaseModel):
    id: int
    goal_id: int
    user_id: int
    title: str
    description: Optional[str] = None
    status: str
    due_date: Optional[date] = None
    sort_order: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ProjectDetailResponse(ProjectResponse):
    tasks: list[TaskInProject] = []

File: web/src/types/project.ts
export type ProjectStatus = 'active' | 'completed' | 'on_hold' | 'archived';

export interface Project {
  id: number;
  goal_id: number;
  user_id: number;
  title: string;
  description: string | null;
  status: ProjectStatus;
  due_date: string | null;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  title: string;
  description?: string;
  status?: ProjectStatus;
  due_date?: string;
  sort_order?: number;
}

export interface ProjectUpdate {
  title?: string;
  description?: string;
  status?: ProjectStatus;
  due_date?: string;
  sort_order?: number;
}

export interface ProjectDetail extends Project {
  tasks?: Array<{
    id: number;
    title: string;
    status: string;
    priority: number;
    due_date: string | null;
    sort_order: number;
    depth: number;
    created_at: string;
  }>;
}

export interface ProjectMember {
  id: number;
  user_id: number;
  display_name: string | null;
  email: string;
  avatar_url: string | null;
  role: 'owner' | 'editor' | 'viewer';
  joined_at: string;
}
